security/experimental_authenticators.rst

```diff
@@ -0,0 +1,434 @@
+    You can then set ``entry_point`` to the FQCN (e.g.
+    ``entry_point: App\Security\CustomEntryPoint``)
+
+.. _authenticators-removed-authentication-providers:
+
+Creating a Custom Authenticator
+-------------------------------
+
+Security traditionally could be extended by writing
+:doc:`custom authentication providers </security/custom_authentication_provider>`.
+The authenticator-based system dropped support for these providers and
+introduced a new authenticator interface as a base for custom
+authentication methods.
+
+.. tip::
+
+    :doc:`Guard authenticators </security/guard_authentication>` are still
+    supported in the authenticator-based system. It is however recommended
+    to also update these when you're refactoring your application to the
+    new system.
+
+    The new authenticator interface has quite some overlap with the guard
+    authenticator interface, making the rewrite easier. To avoid duplicated
+    class names, it is recommended to remove the ``symfony/security-guard``
+    package before rewriting your Security classes:
+
+    .. code-block:: terminal
+
+        $ composer remove symfony/security-guard
+
+Authenticators should implement the
+:class:`Symfony\\Component\\Security\\Http\\Authenticator\\AuthenticatorInterface`.
+You can also extend
+:class:`Symfony\\Component\\Security\\Http\\Authenticator\\AbstractAuthenticator`,
+which has a default implementation for the ``createAuthenticatedToken()``
+method that fits most use-cases::
+
+    // src/Security/ApiKeyAuthenticator.php
+    namespace App\Security;
+
+    use App\Entity\User;
+    use Doctrine\ORM\EntityManagerInterface;
+    use Symfony\Component\HttpFoundation\JsonResponse;
+    use Symfony\Component\HttpFoundation\Request;
+    use Symfony\Component\HttpFoundation\Response;
+    use Symfony\Component\Security\Core\Authentication\Token\TokenInterface;
+    use Symfony\Component\Security\Core\Exception\AuthenticationException;
+    use Symfony\Component\Security\Core\Exception\CustomUserMessageAuthenticationException;
+    use Symfony\Component\Security\Core\Exception\UsernameNotFoundException;
+    use Symfony\Component\Security\Http\Authenticator\AbstractAuthenticator;
+    use Symfony\Component\Security\Http\Authenticator\Passport\PassportInterface;
+    use Symfony\Component\Security\Http\Authenticator\Passport\SelfValidatingPassport;
+
+    class ApiKeyAuthenticator extends AbstractAuthenticator
+    {
+        private $entityManager;
+
+        public function __construct(EntityManagerInterface $entityManager)
+        {
+            $this->entityManager = $entityManager;
+        }
+
+        /**
+         * Called on every request to decide if this authenticator should be
+         * used for the request. Returning `false` will cause this authenticator
+         * to be skipped.
+         *
+         * Returning `null` indicates this authenticator supports to be called
+         * lazily.
```

Comment:
Even for me, I'm not sure what to do with this information :p. Is it relevant for users? If so, what does it *really* mean? When should I return null?

Reply:
It can be very relevant, but let's remove it and discuss it in a cache/lazy firewall related section instead (if we ever write one).